rust-runtime/aws-smithy-types/src/timeout.rs

```diff
@@ -0,0 +1,261 @@
+    connect_timeout: Option<Duration>,
+    tls_negotiation_timeout: Option<Duration>,
+    read_timeout: Option<Duration>,
+    api_call_attempt_timeout: Option<Duration>,
+    api_call_timeout: Option<Duration>,
+}
+
+impl Debug for TimeoutConfig {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            r#"Timeouts:
+Connect (time to first byte):{}
+TLS negotiation:{}
+HTTP read:{}
+API requests:{}
+HTTP requests:{}
+"#,
+            format_timeout(self.connect_timeout),
+            format_timeout(self.tls_negotiation_timeout),
+            format_timeout(self.read_timeout),
+            format_timeout(self.api_call_timeout),
+            format_timeout(self.api_call_attempt_timeout),
+        )
+    }
+}
+
+fn format_timeout(timeout: Option<Duration>) -> String {
+    timeout
+        .map(|d| format!("\t{}s", d.as_secs_f32()))
+        .unwrap_or_else(|| "(unset)".to_owned())
+}
+
+impl TimeoutConfig {
+    /// Create a new `TimeoutConfig` with no timeouts set
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// A limit on the amount of time after making an initial connect attempt on a socket to complete the connect-handshake.
+    pub fn connect_timeout(&self) -> Option<Duration> {
+        self.connect_timeout
+    }
+
+    /// A limit on the amount of time a TLS handshake takes from when the `CLIENT HELLO` message is
+    /// sent to the time the client and server have fully negotiated ciphers and exchanged keys.
+    pub fn tls_negotiation_timeout(&self) -> Option<Duration> {
+        self.tls_negotiation_timeout
+    }
+
+    /// A limit on the amount of time an application takes to attempt to read the first byte over an
+    /// established, open connection after write request. This is also known as the
+    /// "time to first byte" timeout.
+    pub fn read_timeout(&self) -> Option<Duration> {
+        self.read_timeout
+    }
+
+    /// A limit on the amount of time it takes for the first byte to be sent over an established,
+    /// open connection and when the last byte is received from the service for a single attempt.
+    /// If you want to set a timeout for an entire request including retry attempts,
+    /// use [`TimeoutConfig::api_call_timeout`] instead.
+    pub fn api_call_attempt_timeout(&self) -> Option<Duration> {
+        self.api_call_attempt_timeout
+    }
+
+    /// A limit on the amount of time it takes for request to complete. A single request may be
+    /// comprised of several attemps depending on an app's [`RetryConfig`](super::retry::RetryConfig). If you want
+    /// to control timeouts for a single attempt, use [`TimeoutConfig::api_call_attempt_timeout`].
+    pub fn api_call_timeout(&self) -> Option<Duration> {
+        self.api_call_timeout
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the connect timeout
+    pub fn with_connect_timeout(mut self, timeout: Option<Duration>) -> Self {
```

Comment:
Taking a step back and looking at the SDK from a high level, I'm worried we're having too many varied builder patterns across the whole SDK. I'm thinking we should have a tracking issue for going over every builder/config struct in the SDK, establish what patterns we have, choose the ones we like best, and then make them all consistent. Thoughts?

Reply:
I'd love to review our builders. I struggled with what to do with timeouts because the builder and the config were both structs of `Option<Duration>`s and so I ended up removing the builder because it felt redundant and Russell agreed.

Reply:
Tracking in https://github.com/awslabs/smithy-rs/issues/879